src/buffer.rs

```diff
@@ -169,25 +158,32 @@ impl<'source, T: Element> FromPyObject<'source> for PyBuffer<T> {
 impl<T: Element> PyBuffer<T> {
     /// Get the underlying buffer from the specified python object.
     pub fn get(obj: &PyAny) -> PyResult<PyBuffer<T>> {
-        unsafe {
-            let mut buf = Box::pin(ffi::Py_buffer::new());
+        // TODO: use nightly API Box::new_uninit() once stable
+        let mut buf = Box::new(mem::MaybeUninit::uninit());
+        let buf: Box<ffi::Py_buffer> = unsafe {
             err::error_on_minusone(
                 obj.py(),
-                ffi::PyObject_GetBuffer(obj.as_ptr(), &mut *buf, ffi::PyBUF_FULL_RO),
+                ffi::PyObject_GetBuffer(obj.as_ptr(), buf.as_mut_ptr(), ffi::PyBUF_FULL_RO),
             )?;
-            validate(&buf)?;
-            let buf = PyBuffer(buf, PhantomData);
-            // Type Check
-            if mem::size_of::<T>() == buf.item_size()
-                && (buf.0.buf as usize) % mem::align_of::<T>() == 0
-                && T::is_compatible_format(buf.format())
-            {
-                Ok(buf)
-            } else {
-                Err(exceptions::PyBufferError::new_err(
-                    "Incompatible type as buffer",
-                ))
-            }
+            // Safety: buf is initialized by PyObject_GetBuffer.
+            // TODO: use nightly API Box::assume_init() once stable
+            mem::transmute(buf)
+        };
+        // Create PyBuffer immediately so that if validation checks fail, the PyBuffer::drop code
+        // will call PyBuffer_Release (thus avoiding any leaks).
+        let buf = PyBuffer(Pin::from(buf), PhantomData);
+
+        if buf.0.shape.is_null() {
+            Err(PyBufferError::new_err("shape is null but PyBUF_FULL_RO requested"))
+        } else if buf.0.strides.is_null() {
+            Err(PyBufferError::new_err("strides is nullbut PyBUF_FULL_RO requested"))
```

Comment:
Typo:
```suggestion
            Err(PyBufferError::new_err("strides is null but PyBUF_FULL_RO requested"))
```

Also, is mentioning `PyBuF_FULL_R0` ergonomic?

Reply:
> Also, is mentioning PyBuF_FULL_R0 ergonomic?

🤷 I'll take it out